File: ros2_fundamentals/service_server_and_client/src/simple_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "custom_msg_and_srv/srv/add_two_ints.hpp"
//                                |
//          ROS2 convertit AddTwoInts.srv → add_two_ints.hpp (snake_case)

class SimpleServer : public rclcpp::Node
{
public:
    SimpleServer() : Node("simple_server")
    {
        
        // CRÉER LE SERVICE
        server_ = this->create_service<custom_msg_and_srv::srv::AddTwoInts>(
        //                              |
        //                    type du service
            "add_two_ints",   // nom du service
            std::bind(&SimpleServer::serviceCallback, this,
                std::placeholders::_1,   // request
                std::placeholders::_2)   // response
            // 2 placeholders car le callback prend 2 arguments
        );

        RCLCPP_INFO(this->get_logger(), "Server started, waiting for requests...");
    }

private:
    // --------------------------------------------------------
    // CALLBACK DU SERVICE
    // appelé automatiquement quand un client envoie une requête
    // --------------------------------------------------------
    void serviceCallback(
        const custom_msg_and_srv::srv::AddTwoInts::Request::SharedPtr request,
        //    | ce que le client nous envoie (a et b)
        custom_msg_and_srv::srv::AddTwoInts::Response::SharedPtr response)
        //    | ce qu'on va retourner au client (sum)
    {
        // Calculer la somme
        response->sum = request->a + request->b;

        RCLCPP_INFO(this->get_logger(),
            "Request received: a=%ld, b=%ld → sum=%ld",
            request->a, request->b, response->sum);
        // %ld = format pour int64
    }

    rclcpp::Service<custom_msg_and_srv::srv::AddTwoInts>::SharedPtr server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimpleServer>());
    rclcpp::shutdown();
    return 0;
}

File: ros2_fundamentals/service_server_and_client_py/service_server_and_client_py/simple_server.py
import rclpy
from rclpy.node import Node
from custom_msg_and_srv.srv import AddTwoInts

class SimpleServer(Node):
    def __init__(self):
        super().__init__("simple_server")


        # CRÉER LE SERVICE
        self.server_ = self.create_service(
            AddTwoInts,              # type du service
            "add_two_ints",          # nom du service
            self.service_callback    # callback appelé à chaque requête
        )

        self.get_logger().info("Server started, waiting for requests...")

    # CALLBACK DU SERVICE
    def service_callback(self, request, response):
        # request : ce que le client envoie (a et b)
        # response : ce qu'on retourne (sum)

        response.sum = request.a + request.b

        self.get_logger().info(
            f"Request received: a={request.a}, b={request.b} → sum={response.sum}"
        )

        return response   # IMPORTANT : toujours retourner la response !

def main(args=None):
    rclpy.init(args=args)
    node = SimpleServer()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()